sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,652 @@
+      try {
+        if (checkIfStateExists(version - 1)) {
+          found = true
+          lastAvailableVersion = version - 1
+        } else {
+          // Destroy DB so that we can recontruct it using snapshot and delta files
+          RocksDbInstance.destroyDB(rocksDbPath)
+        }
+
+        // Check for snapshot files starting from "version"
+        while (!found && lastAvailableVersion > 0) {
+          found = {
+            try {
+              loadSnapshotFile(lastAvailableVersion)
+            } catch {
+              case e: Exception =>
+                logError(s"$e while reading snapshot file")
+                throw e
+            }
+          }
+          if (!found) {
+            lastAvailableVersion = lastAvailableVersion - 1
+          }
+          logInfo(
+            s"Snapshot for $lastAvailableVersion for " +
+              s"partition ${stateStoreId_.partitionId} found = $found")
+        }
+
+        rocksDbWriteInstance =
+          new OptimisticTransactionDbInstance(keySchema, valueSchema, version.toString)
+        rocksDbWriteInstance.open(rocksDbPath, rocksDbConf)
+        rocksDbWriteInstance.startTransactions()
+
+        // Load all the deltas from the version after the last available
+        // one up to the target version.
+        // The last available version is the one with a full snapshot, so it doesn't need deltas.
+        for (deltaVersion <- (lastAvailableVersion + 1) to version) {
+          val fileToRead = deltaFile(baseDir, deltaVersion)
+          updateFromDeltaFile(
+            fm,
+            fileToRead,
+            keySchema,
+            valueSchema,
+            rocksDbWriteInstance,
+            sparkConf)
+          logInfo(s"Read delta file for version $version of $this from $fileToRead")
+        }
+
+        rocksDbWriteInstance.commit(Some(getBackupPath(version)))
+        rocksDbWriteInstance.close()
+        rocksDbWriteInstance = null
+      } catch {
+        case e: IllegalStateException =>
+          logError(s"Exception while loading state ${e.getMessage}")
+          if (rocksDbWriteInstance != null) {
+            rocksDbWriteInstance.abort()
+            rocksDbWriteInstance.close()
+          }
+          throw e
+      }
+    }
+    logInfo(s"Loading state for $version takes $elapsedMs ms.")
+  }
+
+  private def loadSnapshotFile(version: Long): Boolean = {
+    val fileToRead = snapshotFile(baseDir, version)
+    val tmpLocDir: File = new File(getTempPath(version))
+    val tmpLocFile: File = new File(getTempPath(version).concat(".tar"))
+    try {
+      if (!fm.exists(fileToRead)) {
+        return false
+      }
+      logInfo(s"Will download $fileToRead at location ${tmpLocFile.toString()}")
+      if (downloadFile(fm, fileToRead, new Path(tmpLocFile.getAbsolutePath), sparkConf)) {
+        FileUtility.unTar(tmpLocFile)
+        if (!tmpLocDir.list().exists(_.endsWith(".sst"))) {
+          logWarning("Snaphot files are corrupted")
```

Comment:
This is correct but not exhaustive. No way to indicate the corruption anyway if some sst files are missing.

If we're applying checksum correctly, we might be OK to skip checking this. Do you encounter similar issue while testing/using?